### DISCUSSION TITLE
Pyright is assigning Never type to a clearly defined type

### DISCUSSION THREAD

**Discussion by jmg-duarte** [microsoft/pyright#2495] [Q&A]
I have the following code snippet:

```py
DataT = TypeVar(
    "DataT", bool, int, float, str, List[Any], Dict[str, Any], DataFrame, None
)
class T:
    def f(self, path, type_, data):        
        old_data = self._read_data(path=path, type_=type_) # returns DataT

        if isinstance(data, list):
            if isinstance(old_data, list):
                new_data = old_data + data # old_data: Never
                # Operator "+" not supported for types "Never" and "List[Any]*"
```

I don't get why `old_data` is attributed type `Never`.

I was not able to create a minimal example and the original code is not public.

**Comment by erictraut** [comment]
I tried to create a minimal self-contained example from your code above, but I'm not able to repro the problem. There must be something more involved.

```python
from typing import Any, Dict, List, Type, TypeVar

DataT = TypeVar("DataT", bool, int, float, str, List[Any], Dict[str, Any], None)

def read_data(path, type_: Type[DataT]) -> DataT:
    ...

class T:
    def f(self, path, type_, data):
        old_data = read_data(path=path, type_=type_)

        if isinstance(data, list):
            reveal_type(data) # list[Unknown]
            if isinstance(old_data, list):
                reveal_type(old_data) # list[Unknown]
                new_data = old_data + data
```

Any further thoughts or insights?

**Reply by jmg-duarte** [reply]
Is there any way I can dive in and try to get a better understanding? Like some inference trace?

**Reply by erictraut** [reply]
The best approach is to try to find a minimal reproducible case.

You can use `reveal_type` statements to reveal the evaluated type of an expression at various places in the code. For example, I'm curious the type of `old_data` immediately prior to the `isinstance` conditional.

**Reply by jmg-duarte** [reply]
![Screenshot 2021-10-27 at 21 08 20](https://user-images.githubusercontent.com/15343819/139139378-ab3addc3-8e64-45ef-99da-a91c7d69c6e5.png)

**Reply by erictraut** [reply]
Based on that, the `old_data` is interpreted as an instantiable type (a class), not an object (an instance of a class). You can see that its type is `Type[T]` rather than just `T`. For that reason, `isinstance(old_data, list)` will always be `False`.

I don't have enough visibility into the rest of the code to say whether the type of `old_data` should be an instantiable type at that point, but at least we understand why the type of `old_data` is narrowed to `Never` in this case.
